arch/arm/core/aarch32/CMakeLists.txt

```diff
@@ -34,5 +35,11 @@ add_subdirectory_ifdef(CONFIG_ARM_AARCH32_MMU mmu)
 add_subdirectory_ifdef(CONFIG_CPU_AARCH32_CORTEX_R cortex_a_r)
 add_subdirectory_ifdef(CONFIG_CPU_AARCH32_CORTEX_A cortex_a_r)
 
+if (CONFIG_ARM_ZIMAGE_HEADER)
+zephyr_linker_sources(ROM_START SORT_KEY 0x0vectors zimage_header.ld)
+zephyr_linker_sources(ROM_START SORT_KEY 0x1vectors vector_table.ld)
+zephyr_linker_sources(ROM_START SORT_KEY 0x2vectors cortex_m/vector_table_pad.ld)
+else()
 zephyr_linker_sources(ROM_START SORT_KEY 0x0vectors vector_table.ld)
 zephyr_linker_sources(ROM_START SORT_KEY 0x1vectors cortex_m/vector_table_pad.ld)
+endif()
```

Comment:
What about:
```suggestion
if (CONFIG_ARM_ZIMAGE_HEADER)
zephyr_linker_sources(ROM_START SORT_KEY 0x0vectors zimage_header.ld)
endif()
zephyr_linker_sources(ROM_START SORT_KEY 0x1vectors vector_table.ld)
zephyr_linker_sources(ROM_START SORT_KEY 0x2vectors cortex_m/vector_table_pad.ld)
```

Reply:
@ayankuma ping

Reply:
@carlescufi , all - Apologies for the delay in response, I was very busy for the last few days. :(
I had tried this suggestion before sending my patch, but it did not work for mps2_an385.

I will try this again this week with the latest Zephyr and let you know.

Reply:
I mean, it's not an extremely important fix. I'm mostly surprised that this doesn't work to be honest.

Reply:
@carlocaione should we get this in as-is and then improve it after?

Reply:
> @carlocaione should we get this in as-is and then improve it after?

I don't want to block this (this is why I commented and not blocked) but I'm a bit worried that the fact that my suggestion is not working is hiding some other kind of problems.

Reply:
@carlocaione 
I tried your suggestion and tried executing with mps2_an385

With zImage=y, it works fine. The reason being Cortex M expects the first two entries to be "Initial value for the main stack pointer on reset" and "Start address for the reset handler" (Refer Armv8M DDI0553B.vID16122022, B3.30, Vector tables). I have explained this in commit message

 0:   20001938  
 4:   00001185 

With zImage=n, it does **not** work. The first two entries are incorrect.
0:   00000edd   
4:   00000edd 

The reason being with the following
zephyr_linker_sources(ROM_START SORT_KEY 0x1vectors vector_table.ld)
The vector table now shifts to 0x100, so

 100:   20001938
 104:   00001185

Reply:
I think we can fix that in another PR. Any objections @carlocaione ?

Reply:
nope, no objections.

Reply:
@ayankuma merged this one. Please send a follow-up Pull Request with @carlocaione's suggestion.